doc/adr/0004-indexer.md

```diff
@@ -0,0 +1,123 @@
+New Indexer
+===========
+
+Supporting "mutliple indexes" has been a feature request for some time, but
+why?
+
+- **Workspace Folders**: Supporting multiple workspace folders - particularly
+  useful for editors such as VSCode and Sublime Text.
+- **Index Size**: Being able to reuse f.e. the same `phpstorm-stubs` index on
+  all projects, as currently they are reindexed each time.
+
+Other things:
+
+- Support **multiple files for the same definition**. For example the PHPUnit
+  polyfill `TestCase` will often be located instead of the real PHPUnit
+  `TestCase`, see also the numerous **Laravel doesn't work** issues.
+- **Multiple workers**: Currently all indexing happens in the main Phpactor
+  process. We should instead start a number of indexing **workers** and send
+  files paths (or batches) to them. This would improve indexing
+  performance dramatically.
+
+LSP Workspace Folders
+---------------------
+
+https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_workspaceFolders
+
+Support multiple **project roots** per workspace.
+
+Requirements:
+
+- Need to be able to _add_ or _remove_ workspaces at runtime.
+
+Implementation:
+
+- Workspace folders are ignorant of eachother.
+- Commands such as move class need to apply only on the current "workspace".
+- Workspace symbol search aggregates symbols from _all_ workspace folders.
+
+Scala Metals
+------------
+
+[Worspace Folders Blog Post](https://scalameta.org/metals/blog/2023/07/17/workspace-folders/).
+
+In Metals each endpoint first calls `getServiceFor` which returns an "service"
+configured for the **workspace containing the text document**:
+
+```
+  override def completion(
+      params: CompletionParams
+  ): CompletableFuture[CompletionList] =
+    getServiceFor(params.getTextDocument.getUri).completion(params)
```

Comment:
I think the problem with this approach does not allow cross workspace completions. Which means that would be incompatible with splitting off the stubs into it's own index.